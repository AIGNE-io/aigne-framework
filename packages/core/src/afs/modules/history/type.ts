import type { AFSEntry } from "../../type.js";

export interface HistoryStorageCreateEntry extends Omit<AFSEntry, "id"> {}

export interface HistoryStorageListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export abstract class HistoryStorage {
  abstract create(entry: HistoryStorageCreateEntry): Promise<AFSEntry>;

  abstract list(options?: HistoryStorageListOptions): Promise<{ list: AFSEntry[] }>;

  abstract read(path: string): Promise<AFSEntry | undefined>;
}
