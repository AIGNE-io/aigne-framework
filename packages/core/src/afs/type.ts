import type { Emitter } from "strict-event-emitter";
import type { Context, ContextEventMap } from "../aigne/context.js";
import type { AFSStorage } from "./storage/type.js";

export interface AFSListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  recursive?: boolean;
  maxDepth?: number;
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export interface AFSSearchOptions {
  limit?: number;
}

export interface AFSWriteEntryPayload extends Omit<AFSEntry, "id" | "path"> {}

export interface AFSModule {
  readonly moduleId: string;

  readonly path: string;

  onMount?(root: AFSRoot): void;

  list?(path: string, options?: AFSListOptions): Promise<{ list: AFSEntry[] }>;

  read?(path: string): Promise<AFSEntry | undefined>;

  write?(path: string, content: AFSWriteEntryPayload): Promise<AFSEntry>;

  search?(path: string, query: string, options?: AFSSearchOptions): Promise<{ list: AFSEntry[] }>;
}

export type AFSRootEvents = ContextEventMap & {
  historyCreated: [{ context: Context; entry: AFSEntry }];
};

export interface AFSRoot extends Emitter<AFSRootEvents>, AFSModule {
  storage(module: AFSModule): AFSStorage;
}

export interface AFSEntry<T = any> {
  id: string;
  createdAt?: Date;
  updatedAt?: Date;
  path: string;
  userId?: string | null;
  sessionId?: string | null;
  summary?: string | null;
  metadata?: Record<string, any> | null;
  linkTo?: string | null;
  content?: T;
}
