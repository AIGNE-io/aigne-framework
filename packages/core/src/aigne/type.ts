import type { Agent } from "../agents/agent.ts";

export interface AIGNECLIAgents {
  chat?: Agent;

  agents?: AIGNECLIAgent[];
}

export interface AIGNECLIAgent {
  agent?: Agent;
  name?: string;
  alias?: string[];
  description?: string;
  commands?: AIGNECLIAgent[];
}
