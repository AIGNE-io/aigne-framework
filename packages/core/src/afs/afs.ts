import { Emitter } from "strict-event-emitter";
import { joinURL } from "ufo";
import { logger } from "../utils/logger.js";
import { AFSHistory } from "./modules/history/index.js";
import { SharedAFSStorage, type SharedAFSStorageOptions } from "./storage/index.js";
import type { AFSStorage } from "./storage/type.js";
import type {
  AFSEntry,
  AFSListOptions,
  AFSModule,
  AFSRoot,
  AFSRootEvents,
  AFSSearchOptions,
  AFSWriteEntryPayload,
} from "./type.js";

const DEFAULT_MAX_DEPTH = 5;

export interface AFSOptions {
  storage?: SharedAFSStorage | SharedAFSStorageOptions;
}

export class AFS extends Emitter<AFSRootEvents> implements AFSRoot {
  moduleId: string = "AFSRoot";

  path = "/";

  constructor(options?: AFSOptions) {
    super();

    this._storage =
      options?.storage instanceof SharedAFSStorage
        ? options.storage
        : new SharedAFSStorage(options?.storage);

    this.use(new AFSHistory());
  }

  private _storage: SharedAFSStorage;

  storage(module: AFSModule): AFSStorage {
    return this._storage.withModule(module);
  }

  private modules = new Map<string, AFSModule>();

  use(module: AFSModule) {
    this.modules.set(module.path, module);
    module.onMount?.(this);
    return this;
  }

  async list(path: string, options?: AFSListOptions): Promise<{ list: AFSEntry[] }> {
    const maxDepth = options?.maxDepth ?? DEFAULT_MAX_DEPTH;
    if (!(maxDepth >= 0)) throw new Error(`Invalid maxDepth: ${maxDepth}`);

    const results: AFSEntry[] = [];

    const modules = this.findModules(path);

    for (const { module, subpath, mountPath } of modules) {
      if (!module.list) continue;

      try {
        const newMaxDepth = maxDepth - mountPath.split("/").filter(Boolean).length;
        if (newMaxDepth < 0) continue;

        const { list } = await module.list(subpath, { ...options, maxDepth: newMaxDepth });

        results.push(
          ...list.map((entry) => ({
            ...entry,
            path: joinURL(mountPath, entry.path),
          })),
        );
      } catch (error) {
        logger.error(`Error listing from module at ${mountPath}`, error);
      }
    }

    return { list: results };
  }

  private findModules(
    fullPath: string,
  ): { module: AFSModule; mountPath: string; subpath: string; matchedDepth: number }[] {
    const modules: ReturnType<typeof this.findModules> = [];

    for (const [mountPath, module] of this.modules) {
      const match = this.isSubpath(fullPath, mountPath);
      if (!match) continue;

      modules.push({ ...match, module, mountPath });
    }

    return modules.sort((a, b) => b.matchedDepth - a.matchedDepth);
  }

  private isSubpath(
    fullPath: string,
    mountPath: string,
  ): { subpath: string; matchedDepth: number } | undefined {
    const fullPathSegments = fullPath.split("/").filter(Boolean);
    const mountPathSegments = mountPath.split("/").filter(Boolean);

    const fp = fullPathSegments.join("/");
    const mp = mountPathSegments.join("/");

    if (fp.startsWith(mp)) {
      return {
        matchedDepth: mountPathSegments.length,
        subpath: joinURL("/", ...fullPathSegments.slice(mountPathSegments.length)),
      };
    } else if (mp.startsWith(fp)) {
      return {
        matchedDepth: fullPathSegments.length,
        subpath: joinURL("/", ...mountPathSegments.slice(fullPathSegments.length)),
      };
    }
  }

  async read(path: string): Promise<AFSEntry | undefined> {
    const modules = this.findModules(path);

    for (const { module, mountPath, subpath } of modules) {
      const entry = await module.read?.(subpath);

      if (entry) {
        return {
          ...entry,
          path: joinURL(mountPath, entry.path),
        };
      }
    }
  }

  async write(path: string, content: AFSWriteEntryPayload): Promise<AFSEntry> {
    const module = this.findModules(path)[0];
    if (!module?.module.write) throw new Error(`No module found for path: ${path}`);

    const entry = await module.module.write(module.subpath, content);

    return {
      ...entry,
      path: joinURL(module.mountPath, entry.path),
    };
  }

  async search(
    path: string,
    query: string,
    options?: AFSSearchOptions,
  ): Promise<{ list: AFSEntry[] }> {
    const results: AFSEntry[] = [];

    for (const { module, mountPath, subpath } of this.findModules(path)) {
      if (mountPath.startsWith(path)) {
        if (!module.search) continue;

        try {
          const { list } = await module.search(subpath, query, options);

          results.push(
            ...list.map((entry) => ({
              ...entry,
              path: joinURL(mountPath, entry.path),
            })),
          );
        } catch (error) {
          logger.error(`Error searching in module at ${mountPath}`, error);
        }
      }
    }

    return { list: results };
  }
}
