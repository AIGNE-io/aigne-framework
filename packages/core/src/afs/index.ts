export * from "./afs.js";
export * from "./modules/history/index.js";
export * from "./type.js";
