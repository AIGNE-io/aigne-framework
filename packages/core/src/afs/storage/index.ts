import { and, asc, desc, eq, initDatabase, sql } from "@aigne/sqlite";
import type { AFSEntry, AFSModule } from "../type.js";
import { migrate } from "./migrate.js";
import { entriesTable } from "./models/entries.js";
import type { AFSStorage, AFSStorageCreatePayload, AFSStorageListOptions } from "./type.js";

const DEFAULT_AFS_STORAGE_LIST_LIMIT = 10;

export interface SharedAFSStorageOptions {
  url?: string;
}

export class SharedAFSStorage {
  constructor(public options?: SharedAFSStorageOptions) {}

  private _db: ReturnType<typeof initDatabase> | undefined;

  get db() {
    this._db ??= initDatabase({ url: this.options?.url });

    return this._db;
  }

  withModule(module: AFSModule): AFSStorage {
    return new AFSStorageWithModule(
      this.db,
      migrate(this.db, module).then(() => entriesTable(module)),
    );
  }
}

class AFSStorageWithModule implements AFSStorage {
  constructor(
    private db: ReturnType<typeof initDatabase>,
    private table: Promise<ReturnType<typeof entriesTable>>,
  ) {}

  async list(options: AFSStorageListOptions = {}): Promise<{ list: AFSEntry[] }> {
    const { filter, limit = DEFAULT_AFS_STORAGE_LIST_LIMIT } = options;

    const db = await this.db;
    const table = await this.table;

    const list = await db
      .select()
      .from(table)
      .where(
        and(
          filter?.userId ? eq(table.userId, filter.userId) : undefined,
          filter?.sessionId ? eq(table.sessionId, filter.sessionId) : undefined,
        ),
      )
      .orderBy(
        ...(options.orderBy ?? []).map((item) =>
          (item[1] === "asc" ? asc : desc)(sql.identifier(item[0])),
        ),
      )
      .limit(limit)
      .execute();

    return { list };
  }

  async read(path: string): Promise<AFSEntry | undefined> {
    const db = await this.db;
    const table = await this.table;

    return db
      .select()
      .from(table)
      .where(eq(table.path, path))
      .limit(1)
      .execute()
      .then((memory) => memory.at(0));
  }

  async create(entry: AFSStorageCreatePayload): Promise<AFSEntry> {
    const db = await this.db;
    const table = await this.table;

    let result = await db
      .update(table)
      .set(entry)
      .where(eq(table.path, entry.path))
      .returning()
      .execute();

    if (!result.length) {
      result = await db.insert(table).values(entry).returning().execute();
    }

    const [res] = result;

    if (!res) throw new Error("Failed to create AFS entry, no result");

    return res;
  }
}
