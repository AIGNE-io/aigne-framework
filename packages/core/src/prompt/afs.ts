import type { AFS } from "@aigne/afs";
import { z } from "zod";
import { type Agent, FunctionAgent } from "../agents/agent.js";

export async function getAFSSkills(afs: AFS): Promise<Agent[]> {
  return [
    FunctionAgent.from({
      name: "listModules",
      description: "List all mounted AFS modules",
      process: async () => {
        const modules = await afs.listModules();
        return { modules };
      },
    }),
    FunctionAgent.from({
      name: "list",
      description: "List files in the AFS (AIGNE File System)",
      inputSchema: z.object({
        path: z.string().describe("The path to list files from"),
        options: z
          .object({
            recursive: z.boolean().optional().describe("Whether to list files recursively"),
            maxDepth: z.number().optional().describe("Maximum depth to list files"),
            limit: z.number().optional().describe("Maximum number of entries to return"),
          })
          .optional(),
      }),
      process: async (input) => {
        return afs.list(input.path, input.options);
      },
    }),
    FunctionAgent.from({
      name: "search",
      description: "Search files in the AFS (AIGNE File System)",
      inputSchema: z.object({
        path: z.string().describe("The path to search files in"),
        query: z.string().describe("The search query"),
        options: z
          .object({
            limit: z.number().optional().describe("Maximum number of entries to return"),
          })
          .optional(),
      }),
      process: async (input) => {
        return afs.search(input.path, input.query, input.options);
      },
    }),
    FunctionAgent.from({
      name: "read",
      description: "Read a file from the AFS (AIGNE File System)",
      inputSchema: z.object({
        path: z.string().describe("The path to the file to read"),
      }),
      process: async (input) => {
        const file = await afs.read(input.path);
        return {
          file,
        };
      },
    }),
    FunctionAgent.from({
      name: "write",
      description: "Write a file to the AFS (AIGNE File System)",
      inputSchema: z.object({
        path: z.string().describe("The path to the file to write"),
        content: z.string().describe("The content to write to the file"),
      }),
      process: async (input) => {
        const file = await afs.write(input.path, {
          content: input.content,
        });

        return { file };
      },
    }),
  ];
}
