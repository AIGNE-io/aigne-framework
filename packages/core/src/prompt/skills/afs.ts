import type { AFS } from "@aigne/afs";
import { z } from "zod";
import { type Agent, FunctionAgent } from "../../agents/agent.js";

export async function getAFSSkills(afs: AFS): Promise<Agent[]> {
  return [
    FunctionAgent.from({
      name: "afs_list",
      description:
        "Browse directory contents in the AFS like filesystem ls/tree command - shows files and folders in the specified path",
      inputSchema: z.object({
        path: z.string().describe("The directory path to browse (e.g., '/', '/docs', '/src')"),
        options: z
          .object({
            recursive: z.boolean().optional().describe("Whether to list files recursively"),
            maxDepth: z.number().optional().describe("Maximum depth to list files"),
            limit: z.number().optional().describe("Maximum number of entries to return"),
          })
          .optional(),
      }),
      process: async (input) => {
        const result = await afs.list(input.path, input.options);

        return {
          status: "success",
          tool: "afs_list",
          options: input.options,
          ...result,
        };
      },
    }),
    FunctionAgent.from({
      name: "afs_search",
      description:
        "Find files by searching content using keywords - returns matching files with their paths",
      inputSchema: z.object({
        path: z.string().describe("The directory path to search in (e.g., '/', '/docs')"),
        query: z
          .string()
          .describe(
            "Keywords to search for in file contents (e.g., 'function authentication', 'database config')",
          ),
        options: z
          .object({
            limit: z.number().optional().describe("Maximum number of entries to return"),
          })
          .optional(),
      }),
      process: async (input) => {
        const result = await afs.search(input.path, input.query, input.options);

        return {
          status: "success",
          tool: "afs_search",
          query: input.query,
          options: input.options,
          ...result,
        };
      },
    }),
    FunctionAgent.from({
      name: "afs_read",
      description:
        "Read file contents from the AFS - path must be an exact file path from list or search results",
      inputSchema: z.object({
        path: z
          .string()
          .describe(
            "Exact file path from list or search results (e.g., '/docs/api.md', '/src/utils/helper.js')",
          ),
      }),
      process: async (input) => {
        const result = await afs.read(input.path);

        return {
          status: "success",
          tool: "afs_read",
          path: input.path,
          ...result,
        };
      },
    }),
    FunctionAgent.from({
      name: "afs_write",
      description:
        "Create or update a file in the AFS with new content - overwrites existing files",
      inputSchema: z.object({
        path: z
          .string()
          .describe(
            "Full file path where to write content (e.g., '/docs/new-file.md', '/src/component.js')",
          ),
        content: z.string().describe("The text content to write to the file"),
      }),
      process: async (input) => {
        const result = await afs.write(input.path, {
          content: input.content,
        });

        return {
          status: "success",
          tool: "afs_write",
          path: input.path,
          ...result,
        };
      },
    }),
  ];
}
