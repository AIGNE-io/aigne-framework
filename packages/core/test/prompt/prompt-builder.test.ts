import { expect, spyOn, test } from "bun:test";
import { join } from "node:path";
import { AFS } from "@aigne/afs";
import {
  AIAgent,
  AIAgentToolChoice,
  AIGNE,
  FunctionAgent,
  MCPAgent,
  PromptBuilder,
  TeamAgent,
} from "@aigne/core";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import type { GetPromptResult } from "@modelcontextprotocol/sdk/types.js";
import { z } from "zod";
import { MockMemory } from "../_mocks/mock-memory.js";

test("PromptBuilder should build messages correctly", async () => {
  const context = new AIGNE().newContext();

  const builder = PromptBuilder.from("Test instructions");

  const memory = new MockMemory({});
  await memory.record(
    {
      content: [{ input: { message: "Hello, How can I help you?" }, source: "TestAgent" }],
    },

    context,
  );

  const agent = AIAgent.from({
    memory,
    inputKey: "message",
  });

  const prompt1 = await builder.build({
    agent,
    input: { message: "Hello" },
    context,
  });

  expect(prompt1.messages).toEqual([
    {
      role: "system",
      content: "Test instructions",
    },
    {
      role: "system",
      content: expect.stringContaining("Hello, How can I help you?"),
    },
    {
      role: "user",
      content: [{ type: "text", text: "Hello" }],
    },
  ]);

  const prompt2 = await builder.build({
    input: { name: "foo" },
    context,
  });
  expect(prompt2.messages).toEqual([
    {
      role: "system",
      content: "Test instructions",
    },
  ]);
});

test("PromptBuilder should build response format correctly", async () => {
  const context = new AIGNE().newContext();

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    outputSchema: z.object({
      name: z.string(),
      age: z.number().optional(),
    }),
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.responseFormat).toEqual({
    type: "json_schema",
    jsonSchema: {
      name: "output",
      schema: expect.objectContaining({
        type: "object",
        properties: {
          name: { type: "string" },
          age: { type: "number" },
        },
        required: ["name"],
        additionalProperties: false,
      }),
      strict: true,
    },
  });
});

test("PromptBuilder should build skills correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
    inputSchema: z.object({
      name: z.string(),
      age: z.number().optional(),
    }),
  });

  const teamAgentSkill = TeamAgent.from({
    name: "TestTeamAgent",
    skills: [
      AIAgent.from({
        name: "TestSkillInTeamAgent1",
      }),
      AIAgent.from({
        name: "TestSkillInTeamAgent2",
      }),
    ],
  });

  // MCPAgent is not invokable, so we use it as a skill container
  const mcpAgentSkill = new MCPAgent({
    name: "TestMCPAgent",
    client: new Client({
      name: "TestClient",
      version: "1.0.0",
    }),
    skills: [
      AIAgent.from({
        name: "TestMCPSkill1",
        instructions: "TestMCPSkill1 instructions",
      }),
      AIAgent.from({
        name: "TestMCPSkill2",
        instructions: "TestMCPSkill2 instructions",
      }),
    ],
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill, teamAgentSkill, mcpAgentSkill],
    toolChoice: skill,
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.tools).toEqual([
    {
      type: "function",
      function: {
        name: "TestSkill",
        description: "Test skill description",
        parameters: expect.objectContaining({
          type: "object",
          properties: {
            name: { type: "string" },
            age: { type: "number" },
          },
          required: ["name"],
          additionalProperties: false,
        }),
      },
    },
    // TeamAgent as skill, nested skills should no longer be included in the tools
    {
      type: "function",
      function: {
        name: "TestTeamAgent",
        parameters: {},
      },
    },
    // MCPAgent is not invokable, so it should not be included in the tools, and the nested skills should be included
    {
      type: "function",
      function: {
        name: "TestMCPSkill1",
        parameters: {},
      },
    },
    {
      type: "function",
      function: {
        name: "TestMCPSkill2",
        parameters: {},
      },
    },
  ]);

  expect(prompt.toolChoice).toEqual({
    type: "function",
    function: {
      name: skill.name,
      description: skill.description,
    },
  });
});

test("PromptBuilder should unique skills correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill, skill],
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.tools).toHaveLength(1);
});

test("PromptBuilder should build toolChoice with router mode correctly", async () => {
  const context = new AIGNE().newContext();

  const skill = FunctionAgent.from({
    name: "TestSkill",
    description: "Test skill description",
    process: () => ({}),
  });

  const agent = AIAgent.from({
    name: "TestAgent",
    instructions: "Test instructions",
    skills: [skill],
    toolChoice: AIAgentToolChoice.router,
  });

  const prompt = await agent.instructions.build({ input: {}, agent, context });

  expect(prompt.toolChoice).toEqual("required");
});

test("PromptBuilder from string", async () => {
  const context = new AIGNE().newContext();

  const builder = PromptBuilder.from("Hello, {{agentName}}!");

  const prompt = await builder.build({ input: { agentName: "Alice" }, context });

  expect(prompt).toEqual({
    messages: [
      {
        role: "system",
        content: "Hello, Alice!",
      },
    ],
  });
});

test("PromptBuilder from MCP prompt result", async () => {
  const context = new AIGNE().newContext();

  const prompt: GetPromptResult = {
    description: "Test prompt",
    messages: [
      {
        role: "user",
        content: {
          type: "text",
          text: "Hello",
        },
      },
      {
        role: "user",
        content: {
          type: "resource",
          resource: {
            uri: "echo://Hello",
            text: "Resource echo: Hello",
            mimeType: "text/plain",
          },
        },
      },
      {
        role: "user",
        content: {
          type: "resource",
          resource: {
            uri: "test://image-resource",
            blob: Buffer.from("FAKE IMAGE FROM RESOURCE").toString("base64"),
            mimeType: "image/png",
          },
        },
      },
      {
        role: "user",
        content: {
          type: "image",
          mimeType: "image/png",
          data: Buffer.from("FAKE IMAGE").toString("base64"),
        },
      },
      {
        role: "assistant",
        content: {
          type: "text",
          text: "How can I help you?",
        },
      },
    ],
  };

  const promptBuilder = PromptBuilder.from(prompt);
  expect(await promptBuilder.build({ context })).toEqual(
    expect.objectContaining({
      messages: [
        {
          role: "user",
          content: "Hello",
        },
        {
          role: "user",
          content: "Resource echo: Hello",
        },
        {
          role: "user",
          content: [
            { type: "url", url: Buffer.from("FAKE IMAGE FROM RESOURCE").toString("base64") },
          ],
        },
        {
          role: "user",
          content: [{ type: "url", url: Buffer.from("FAKE IMAGE").toString("base64") }],
        },
        {
          role: "agent",
          content: "How can I help you?",
        },
      ],
    }),
  );
});

test("PromptBuilder from file", async () => {
  const context = new AIGNE().newContext();

  const path = join(import.meta.dirname, "test-prompt.txt");

  const builder = PromptBuilder.from({ path });

  const prompt = await builder.build({
    input: { agentName: "Alice", language: "English" },
    context,
  });

  expect(prompt).toMatchSnapshot();
});

test("PromptBuilder should build image prompt correctly", async () => {
  const builder = PromptBuilder.from("Draw an image about {{topic}}");

  expect(await builder.buildImagePrompt({ input: { topic: "a cat" } })).toEqual({
    prompt: "Draw an image about a cat",
  });
});

test("PromptBuilder should build with afs correctly", async () => {
  const builder = PromptBuilder.from("Test instructions");

  const afs = new AFS();

  const agent = AIAgent.from({
    inputKey: "message",
    afs,
    afsConfig: {
      injectHistory: false,
      historyWindowSize: 5,
    },
  });

  // Build without AFS history
  const result = await builder.build({
    agent,
  });

  expect(result.messages).toMatchInlineSnapshot(`
    [
      {
        "content": "Test instructions",
        "name": undefined,
        "role": "system",
      },
      SystemMessageTemplate {
        "content": 
    "
    <afs_usage>
    AFS (AIGNE File System) provides tools to interact with a virtual file system, allowing you to list, search, read, and write files. Use these tools to manage and retrieve files as needed.

    Modules:
    - moduleId: AFSHistory
      path: /history


    Available Tools:
    1. afs_list: Browse directory contents like filesystem ls/tree command - shows files and folders in a given path
    2. afs_search: Find files by content keywords - use specific keywords related to what you're looking for
    3. afs_read: Read file contents - path must be an exact file path from list or search results
    4. afs_write: Write content to a file in the AFS

    Workflow: Use afs_list to browse directories, afs_search to find specific content, then afs_read to access file contents.
    </afs_usage>
    "
    ,
        "name": undefined,
        "options": undefined,
        "role": "system",
      },
    ]
  `);

  // Mock AFS history
  agent.afsConfig = { ...agent.afsConfig, injectHistory: true };

  const listSpy = spyOn(afs, "list").mockResolvedValueOnce({
    list: [
      {
        id: "1",
        path: "/history/1",
        content: { input: { message: "Hello" }, output: { message: "Hello, How can I help you?" } },
      },
      {
        id: "1",
        path: "/history/1",
        content: { input: { message: "I'm Bob" }, output: { message: "Hello, Bob!" } },
      },
    ],
  });

  const result1 = await builder.build({
    agent,
  });

  expect(listSpy.mock.calls).toMatchInlineSnapshot(`
    [
      [
        "/history",
        {
          "limit": 5,
          "orderBy": [
            [
              "createdAt",
              "desc",
            ],
          ],
        },
      ],
    ]
  `);

  expect(result1).toMatchInlineSnapshot(
    { toolAgents: expect.anything() },
    `
    {
      "messages": [
        {
          "content": "Test instructions",
          "name": undefined,
          "role": "system",
        },
        SystemMessageTemplate {
          "content": 
    "
    <afs_usage>
    AFS (AIGNE File System) provides tools to interact with a virtual file system, allowing you to list, search, read, and write files. Use these tools to manage and retrieve files as needed.

    Modules:
    - moduleId: AFSHistory
      path: /history


    Available Tools:
    1. afs_list: Browse directory contents like filesystem ls/tree command - shows files and folders in a given path
    2. afs_search: Find files by content keywords - use specific keywords related to what you're looking for
    3. afs_read: Read file contents - path must be an exact file path from list or search results
    4. afs_write: Write content to a file in the AFS

    Workflow: Use afs_list to browse directories, afs_search to find specific content, then afs_read to access file contents.
    </afs_usage>
    "
    ,
          "name": undefined,
          "options": undefined,
          "role": "system",
        },
        {
          "content": [
            {
              "text": "Hello",
              "type": "text",
            },
          ],
          "role": "user",
        },
        {
          "content": [
            {
              "text": "Hello, How can I help you?",
              "type": "text",
            },
          ],
          "role": "agent",
        },
        {
          "content": [
            {
              "text": "I'm Bob",
              "type": "text",
            },
          ],
          "role": "user",
        },
        {
          "content": [
            {
              "text": "Hello, Bob!",
              "type": "text",
            },
          ],
          "role": "agent",
        },
      ],
      "modelOptions": undefined,
      "outputFileType": undefined,
      "responseFormat": undefined,
      "toolAgents": Anything,
      "toolChoice": "auto",
      "tools": [
        {
          "function": {
            "description": "Browse directory contents in the AFS like filesystem ls/tree command - shows files and folders in the specified path",
            "name": "afs_list",
            "parameters": {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": false,
              "properties": {
                "options": {
                  "additionalProperties": false,
                  "properties": {
                    "limit": {
                      "description": "Maximum number of entries to return",
                      "type": "number",
                    },
                    "maxDepth": {
                      "description": "Maximum depth to list files",
                      "type": "number",
                    },
                    "recursive": {
                      "description": "Whether to list files recursively",
                      "type": "boolean",
                    },
                  },
                  "required": [],
                  "type": "object",
                },
                "path": {
                  "description": "The directory path to browse (e.g., '/', '/docs', '/src')",
                  "type": "string",
                },
              },
              "required": [
                "path",
              ],
              "type": "object",
            },
          },
          "type": "function",
        },
        {
          "function": {
            "description": "Find files by searching content using keywords - returns matching files with their paths",
            "name": "afs_search",
            "parameters": {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": false,
              "properties": {
                "options": {
                  "additionalProperties": false,
                  "properties": {
                    "limit": {
                      "description": "Maximum number of entries to return",
                      "type": "number",
                    },
                  },
                  "required": [],
                  "type": "object",
                },
                "path": {
                  "description": "The directory path to search in (e.g., '/', '/docs')",
                  "type": "string",
                },
                "query": {
                  "description": "Keywords to search for in file contents (e.g., 'function authentication', 'database config')",
                  "type": "string",
                },
              },
              "required": [
                "path",
                "query",
              ],
              "type": "object",
            },
          },
          "type": "function",
        },
        {
          "function": {
            "description": "Read file contents from the AFS - path must be an exact file path from list or search results",
            "name": "afs_read",
            "parameters": {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": false,
              "properties": {
                "path": {
                  "description": "Exact file path from list or search results (e.g., '/docs/api.md', '/src/utils/helper.js')",
                  "type": "string",
                },
              },
              "required": [
                "path",
              ],
              "type": "object",
            },
          },
          "type": "function",
        },
        {
          "function": {
            "description": "Create or update a file in the AFS with new content - overwrites existing files",
            "name": "afs_write",
            "parameters": {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": false,
              "properties": {
                "content": {
                  "description": "The text content to write to the file",
                  "type": "string",
                },
                "path": {
                  "description": "Full file path where to write content (e.g., '/docs/new-file.md', '/src/component.js')",
                  "type": "string",
                },
              },
              "required": [
                "path",
                "content",
              ],
              "type": "object",
            },
          },
          "type": "function",
        },
      ],
    }
  `,
  );
});
