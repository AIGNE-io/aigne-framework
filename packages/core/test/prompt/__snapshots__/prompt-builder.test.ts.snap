// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptBuilder from file 1`] = `
{
  "messages": [
    {
      "content": 
"## instructions

You are Alice, a professional customer service.

Output in English

"
,
      "name": undefined,
      "role": "system",
    },
  ],
  "modelOptions": undefined,
  "responseFormat": undefined,
  "toolAgents": undefined,
  "toolChoice": undefined,
  "tools": undefined,
}
`;
